### PR TITLE
Add USD plugin

### PR DESCRIPTION
**Pull request by Meakk**

### CODE REVIEW

File: plugins/usd/configs/thumbnail.d/10_usd.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  ".*(usd)[acz]?":
+  {
+    "load-plugins": "alembic"
```
Comment (mwestphal) [inline plugins/usd/configs/thumbnail.d/10_usd.json:4]: nope

File: testing/data/suzanne.usd
Comment (mwestphal) [inline testing/data/suzanne.usd:1]: data source ?

File: plugins/usd/f3d-usd-formats.xml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<mime-info xmlns="http://www.freedesktop.org/standards/shared-mime-info">
+  <mime-type type="model/vnd.usda">
+    <comment>USD ASCII</comment>
+    <glob pattern="*.usda"/>
+  </mime-type>
+  <mime-type type="model/vnd.usdz+zip">
+    <comment>Compressed USD</comment>
+    <glob pattern="*.usdz"/>
```
[the remark this answers is not in the document]
Comment (Meakk) [reply plugins/usd/f3d-usd-formats.xml:9]: I couldn't find any info related to usdc mimetype.  
USDC is the binary version of USD but it hasn't been registered apparently.

[the remark this answers is not in the document]
Comment (mwestphal) [reply plugins/usd/f3d-usd-formats.xml:9]: Looks like it is indeed missing from iana, just add it like the other ones, with the `vnd` style ?

File: testing/data/suzanne.usd
Comment (Meakk) [reply testing/data/suzanne.usd:1]: created myself with Blender USD exporter and the regular suzanne model.

File: plugins/usd/f3d-usd-formats.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Meakk) [reply plugins/usd/f3d-usd-formats.xml:17]: ok, I've added USD and USDC.

**Comment by mwestphal** [comment]
What is blocking this @Meakk ?

File: .github/actions/boost-install-dep/action.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
+name: 'Install Boost Dependency'
+description: 'Install Boost Dependency using cache when possible'
```
Comment (mwestphal) [inline .github/actions/boost-install-dep/action.yml:2]: did you try installing boost in the sb ? Does it work ?

File: .github/actions/generic-ci/action.yml [hunk 4]
```diff
@@ -122,7 +122,7 @@ runs:
       if: runner.os == 'Windows'
       shell: powershell
       working-directory: ${{github.workspace}}
-      run: echo "PATH=$env:PATH;$(pwd)\dependencies\install\bin\;$(pwd)\install\bin\;"| Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
+      run: echo "PATH=$env:PATH;$(pwd)\dependencies\install\bin\;$(pwd)\dependencies\install\lib\;$(pwd)\install\bin\;"| Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
```
Comment (mwestphal) [inline .github/actions/generic-ci/action.yml:125]: can we fix that in boost or usd install instead ?

File: .github/actions/lfs-copy/action.yml [hunk 5]
```diff
@@ -22,7 +22,7 @@ runs:
       uses: actions/cache@v3
       with:
         path: lfs_data
-        key: lfs-data-${{inputs.lfs_sha}}-1
+        key: lfs-data-${{inputs.lfs_sha}}-13
```
Comment (mwestphal) [inline .github/actions/lfs-copy/action.yml:25]: can you loop to 0 instead ?

File: .github/actions/usd-install-dep/0001-usd-onetbb.patch
Comment (mwestphal) [inline .github/actions/usd-install-dep/0001-usd-onetbb.patch:1]: Ill not review this patch, please add a comment somewhere explaining where this patch is coming from

File: plugins/usd/f3d-usd-formats.xml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<mime-info xmlns="http://www.freedesktop.org/standards/shared-mime-info">
+  <mime-type type="model/vnd.usd">
```
Comment (mwestphal) [inline plugins/usd/f3d-usd-formats.xml:3]: not coherent with what is declared in the CMakeLists.txt. Only what is not available in IANA should be listed in this file.

**Comment by mwestphal** [comment]
Half of the reader code is not covered, we need to fix that :)

File: .github/actions/boost-install-dep/action.yml (hunk 3 above)
Comment (Meakk) [reply .github/actions/boost-install-dep/action.yml:2]: there's already a boost project in the common-sb

File: .github/actions/generic-ci/action.yml (hunk 4 above)
Comment (Meakk) [reply .github/actions/generic-ci/action.yml:125]: Why would we want to fix it? It's legit, "modules" (dynamically loaded libs) on Windows are installed in `lib`, not `bin`

File: plugins/usd/configs/config.d/10_usd.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
+  ".*(usd)[acz]?":
+  {
+    "load-plugins": "usd"
```
[the remark this answers is not in the document]
Comment (Meakk) [reply plugins/usd/configs/config.d/10_usd.json:4]: No. The up direction for example is specified within the file...  
I think it would make sense for full scenes to configure the up direction in the importer at runtime.  
But I won't do that in this PR.

File: .github/actions/generic-ci/action.yml (hunk 4 above)
Comment (mwestphal) [reply .github/actions/generic-ci/action.yml:125]: dlls are expected to be in bin, not lib.

File: .github/actions/boost-install-dep/action.yml (hunk 3 above)
Comment (mwestphal) [reply .github/actions/boost-install-dep/action.yml:2]: yes, but did you try it ? Did you try a usd project as well ? Just checking how hard it will be to deliver this in the 2.2 binary release.

File: plugins/usd/f3d-usd-formats.xml (hunk 6 above)
Comment (Meakk) [reply plugins/usd/f3d-usd-formats.xml:3]: I'm confused. What should I do?  
See https://github.com/f3d-app/f3d/pull/819#discussion_r1218291379

File: .github/actions/boost-install-dep/action.yml (hunk 3 above)
Comment (Meakk) [reply .github/actions/boost-install-dep/action.yml:2]: I guess we'll see? If that's too complicated, we'll drop it.

File: .github/actions/generic-ci/action.yml (hunk 4 above)
Comment (Meakk) [reply .github/actions/generic-ci/action.yml:125]: No. DLLs are in `bin` if the libraries are `SHARED`, but in `lib` if the libraries are `MODULE`.

Comment (Meakk) [reply .github/actions/generic-ci/action.yml:125]: See https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#library-output-artifacts

Comment (mwestphal) [reply .github/actions/generic-ci/action.yml:125]: https://stackoverflow.com/questions/4845984/difference-between-modules-and-shared-libraries

Comment (mwestphal) [reply .github/actions/generic-ci/action.yml:125]: Any idea why we are getting a module ?

Why does it matter to put in the PATH then ?

File: plugins/usd/f3d-usd-formats.xml (hunk 6 above)
Comment (mwestphal) [reply plugins/usd/f3d-usd-formats.xml:3]: these .xml files are used to augment the mime types database on linux. If a format is already declared in IANA, there is great chances that it is also already present in the freedesktop database of mime types, which can easily be checked locally.

If a format is declared in IANA but not present in the freedesktop database, then it should be declared here with the right IANA name, it could also be contributed back to freedesktop upstream if we want to.

If a format is not declared in IANA (and objvously not present in the freedesktop database), then it should be declared here  with the `application/vnd.ext` style.

Once a format is correctly declared in the mimetypes database, it can be used by the thumbnailers, if the mime-type are correct.


In your specific case, afaiu:
 - IANA describes `model/vnd.usda` and `model/vnd.usdz`
 - Freedesktop doesnt have any of these mimetypes
 
 So what you need to do:
  - In this .xml file, declare `model/vnd.usda` and `model/vnd.usdz`
  - In this .xml file, declare `application/vnd.usd` and `application/vnd.usdc`
  - in the thumbnail file, use the right mimetype described in the xml file
  - Optionally, contribute back `model/vnd.usda` and `model/vnd.usdz` to freedesktop

File: plugins/usd/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+include(GNUInstallDirs)
+
+# Check if the plugin is built externally
+if(CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
+  find_package(f3d REQUIRED)
+else()
+  include(${CMAKE_SOURCE_DIR}/cmake/f3dPlugin.cmake)
+endif()
+
+find_package(pxr REQUIRED)
```
Comment (mwestphal) [inline plugins/usd/CMakeLists.txt:14]: what is pxr ?

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 9]
```diff
@@ -0,0 +1,953 @@
...
+#include <pxr/usd/usdGeom/tokens.h>
+#include <pxr/usd/usdGeom/xform.h>
+#include <pxr/usd/usdShade/material.h>
+#include <pxr/usd/usdShade/materialBindingAPI.h>
+#include <pxr/usd/usdSkel/bakeSkinning.h>
+#if defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+
+using namespace pxr;
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:60]: lets not ?

File: plugins/usd/CMakeLists.txt (hunk 8 above)
Comment (Meakk) [reply plugins/usd/CMakeLists.txt:14]: "pixar" I guess?

**Comment by mwestphal** [comment]
@Meakk could we move this one forward with SSIM or PSNR image comparison ?

**Comment by Meakk** [comment]
I just need to add another test for covering the remaining lines. Feel free to review again.

File: plugins/usd/module/vtkF3DUSDImporter.cxx (hunk 9 above)
Comment (mwestphal) [reply plugins/usd/module/vtkF3DUSDImporter.cxx:60]: lets not use using

File: plugins/usd/CMakeLists.txt (hunk 8 above)
Comment (mwestphal) [reply plugins/usd/CMakeLists.txt:14]: so we depend on pxr ? does usd depend on it ?

File: .github/actions/generic-ci/action.yml (hunk 4 above)
Comment (mwestphal) [reply .github/actions/generic-ci/action.yml:125]: still interested by the answer here

File: plugins/usd/module/vtkF3DUSDImporter.h [hunk 10]
```diff
@@ -0,0 +1,72 @@
+/**
+ * @class   vtkF3DUSDImporter
+ * @brief   Importer using USD library
+ *
+ * TODO
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.h:5]: TODO

**Review by mwestphal — CHANGES_REQUESTED** [review]
lots of not yet covered code

File: .github/actions/generic-ci/action.yml (hunk 4 above)
Comment (Meakk) [reply .github/actions/generic-ci/action.yml:125]: A module on Windows is a DLL loaded dynamically and has no `.lib` associated import file. It's intended to be loaded by the `LoadLibrary` API. We have to put it in the PATH because Windows is looking in it when calling `LoadLibrary`.

File: plugins/usd/CMakeLists.txt (hunk 8 above)
Comment (Meakk) [reply plugins/usd/CMakeLists.txt:14]: USD package is pxr. I know it's weird but that's how the CMake package is called.

Comment (mwestphal) [reply plugins/usd/CMakeLists.txt:14]: It all makes sense now ^^

File: .github/actions/generic-ci/action.yml (hunk 4 above)
Comment (mwestphal) [reply .github/actions/generic-ci/action.yml:125]: ok got it.

File: plugins/usd/f3d-usd-formats.xml (hunk 6 above)
Comment (Meakk) [reply plugins/usd/f3d-usd-formats.xml:3]: I see, thank you

Comment (mwestphal) [reply plugins/usd/f3d-usd-formats.xml]: I'd say yes but I'm not enterely sure. Just check the thumbnailers works locally ?

**Comment by Meakk** [comment]
@mwestphal please take a look. Coverage is good now and all comments have been addressed.

File: .github/actions/lfs-copy/action.yml [hunk 11]
```diff
@@ -22,7 +22,7 @@ runs:
       uses: actions/cache@v3
       with:
         path: lfs_data
-        key: lfs-data-${{inputs.lfs_sha}}-1
+        key: lfs-data-${{inputs.lfs_sha}}-0
```
Comment (mwestphal) [inline .github/actions/lfs-copy/action.yml:25]: ok to reset to 0 but I dont see why this change would be needed

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 12]
```diff
@@ -0,0 +1,966 @@
...
+
+        auto mat = this->GetLocalTransform(xform, timeCode);
+
+        this->ImportNode(renderer, prim.GetPrototype(), path.AppendChild(prim.GetName()), mat);
+      }
+      else if (prim.IsA<pxr::UsdGeomPointInstancer>())
+      {
+        pxr::UsdGeomPointInstancer glyphs = pxr::UsdGeomPointInstancer(prim);
+
+        // Ideally, we should use the 3D glyph mapper, but it's left for future work
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:124]: Add a TODO ?

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 13]
```diff
@@ -0,0 +1,966 @@
...
+    int maxWidth = 0;
+    int maxHeight = 0;
+
+    if (occlusionImage)
+    {
+      int* size = occlusionImage->GetDimensions();
+
+      maxWidth = std::max(maxWidth, size[0]);
+      maxHeight = std::max(maxHeight, size[1]);
+    }
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:545]: occlusionImage is not covered.

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 14]
```diff
@@ -0,0 +1,966 @@
...
+
+    if (tex == nullptr)
+    {
+      pxr::TfToken idToken;
+      bool defined = samplerPrim.GetIdAttr().Get(&idToken);
+      if (!defined || idToken != pxr::TfToken("UsdUVTexture"))
+      {
+        // only pxr::UsdUVTexture supported for now
+        return nullptr;
+      }
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:629]: should be easy to test with an unsuported texture type ?

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 15]
```diff
@@ -0,0 +1,966 @@
...
+    }
+    else if (channels == "a")
+    {
+      extract->SetComponents(3);
+    }
+    else
+    {
+      // just return the image as is
+      return tex;
+    }
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:704]: can this actually happen with a valid file ? If so we should be able to test it.

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 16]
```diff
@@ -0,0 +1,966 @@
...
+      {
+        pxr::TfToken materialToken;
+
+        bool defined = attr.Get(&materialToken);
+
+        if (!defined || materialToken != pxr::TfToken("UsdPreviewSurface"))
+        {
+          // only pxr::UsdPreviewSurface supported for now
+          return { nullptr, isTranslucent };
+        }
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:731]: same, should be testable

File: plugins/usd/module/vtkF3DUSDImporter.h [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+  void EnableAnimation(vtkIdType vtkNotUsed(animationIndex)) override
+  {
+      this->AnimationEnabled = true;
+  }
+
+  void DisableAnimation(vtkIdType vtkNotUsed(animationIndex)) override
+  {
+    this->AnimationEnabled = false;
+  }
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.h:55]: should be trivial to cover in the Cxx test

File: testing/data/DATA_LICENSES.md [hunk 18]
```diff
@@ -3,6 +3,8 @@
 # Copyrights and Licenses
 
 - albedo.png: glTF-Sample-Models: Public Domain
+- AnimatedCube.usdz: [USD Working Group](https://github.com/usd-wg/assets): Apache 2.0
+- McUSD.usdz: [USD Working Group](https://github.com/usd-wg/assets): CC-NC-BY-SA, textures by [jasonjgardner](https://github.com/jasonjgardner)
```
Comment (mwestphal) [inline testing/data/DATA_LICENSES.md:7]: way more data were added than two :)

**Review by mwestphal — CHANGES_REQUESTED** [review]
Missing a https://f3d.app doc update

File: .github/actions/lfs-copy/action.yml (hunk 11 above)
Comment (Meakk) [reply .github/actions/lfs-copy/action.yml:25]: it's not, I'll revert

File: plugins/usd/module/vtkF3DUSDImporter.cxx (hunk 14 above)
Comment (Meakk) [reply plugins/usd/module/vtkF3DUSDImporter.cxx:629]: I think it's the most common texture type, if not the only one. The library is designed for future evolution so there are a lot of "token" like that.

File: plugins/usd/module/vtkF3DUSDImporter.cxx (hunk 15 above)
Comment (Meakk) [reply plugins/usd/module/vtkF3DUSDImporter.cxx:704]: Maybe? Some exotic `rb` dual channel? I don't know but certainly no trivial to test.

File: plugins/usd/module/vtkF3DUSDImporter.cxx (hunk 16 above)
Comment (Meakk) [reply plugins/usd/module/vtkF3DUSDImporter.cxx:731]: Couldn't find an asset with a different material token.

File: testing/data/DATA_LICENSES.md (hunk 18 above)
Comment (Meakk) [reply testing/data/DATA_LICENSES.md:7]: all other one are my own creation

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 19]
```diff
@@ -0,0 +1,966 @@
...
+            polydata->SetPoints(points);
+          }
+
+          // read normals
+          pxr::UsdAttribute normalsAttr = meshPrim.GetNormalsAttr();
+
+          if (meshPrim.GetNormalsInterpolation() == pxr::TfToken("vertex"))
+          {
+            // only vertex interpolation is supported for now
+            // there are many datasets using faceVarying to represent sharp edges
```
Comment (Meakk) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:185]: @mwestphal FYI I'm planning to drop support of face-varying for USD in this PR. I'll log an issue and ideally we need to create a VTKExtension filter to factorize the code between Alembic and USD.

File: README.md [hunk 20]
```diff
@@ -69,7 +69,7 @@ We envision F3D as a community-driven project, we love to see how the project is
 
 # Acknowledgments
 
-F3D was initially created by [Kitware SAS](https://www.kitware.eu/) and is relying on many awesome open source projects, including [VTK](https://vtk.org/), [OCCT](https://dev.opencascade.org/), [Assimp](https://www.assimp.org/), [Alembic](http://www.alembic.io/), [Draco](https://google.github.io/draco/) and [OSPRay](https://www.ospray.org/).
+F3D was initially created by [Kitware SAS](https://www.kitware.eu/) and is relying on many awesome open source projects, including [VTK](https://vtk.org/), [OCCT](https://dev.opencascade.org/), [Assimp](https://www.assimp.org/), [Alembic](http://www.alembic.io/), [Draco](https://google.github.io/draco/), [OpenUSD](https://openusd.org/release/index.html) and [OSPRay](https://www.ospray.org/).
```
Comment (mwestphal) [inline README.md:72]: Add usd in the list of "premium" extensions

File: application/testing/CMakeLists.txt [hunk 21]
```diff
@@ -553,6 +553,28 @@ if(F3D_PLUGIN_BUILD_OCCT)
...
+if(F3D_PLUGIN_BUILD_USD)
+  f3d_test(NAME TestUSD DATA suzanne.usd ARGS --load-plugins=usd --up=+Z DEFAULT_LIGHTS)
+  f3d_test(NAME TestUSDAPrimitives DATA primitives.usda ARGS --load-plugins=usd DEFAULT_LIGHTS)
+  f3d_test(NAME TestUSDAPrimitivesZAxis DATA primitivesZ.usda ARGS --load-plugins=usd DEFAULT_LIGHTS)
+  f3d_test(NAME TestUSDAInstancing DATA instancing.usda ARGS --load-plugins=usd DEFAULT_LIGHTS)
+  f3d_test(NAME TestUSDAGlyphs DATA glyphs.usda ARGS --load-plugins=usd DEFAULT_LIGHTS)
+  f3d_test(NAME TestUSDInvalid DATA invalid.usd REGEXP "Failed to open layer" ARGS --load-plugins=usd NO_BASELINE)
+
+  # This test is there to test occlusion texture and face-varying point data
+  # TODO: Note that the result looks incorrect because of face-varying attributes and must be fixed later
```
Comment (mwestphal) [inline application/testing/CMakeLists.txt:565]: please create a f3d issue

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 22]
```diff
@@ -0,0 +1,968 @@
...
+
+        auto mat = this->GetLocalTransform(xform, timeCode);
+
+        this->ImportNode(renderer, prim.GetPrototype(), path.AppendChild(prim.GetName()), mat);
+      }
+      else if (prim.IsA<pxr::UsdGeomPointInstancer>())
+      {
+        pxr::UsdGeomPointInstancer glyphs = pxr::UsdGeomPointInstancer(prim);
+
+        // TODO: Ideally, we should use the 3D glyph mapper, but it's left for future work
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:124]: create an issue ?

File: plugins/usd/module/vtkF3DUSDImporter.cxx (hunk 19 above)
Comment (mwestphal) [reply plugins/usd/module/vtkF3DUSDImporter.cxx:169]: please create an issue

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 23]
```diff
@@ -0,0 +1,968 @@
...
+  void ImportRoot(vtkRenderer* renderer)
+  {
+    if (!this->Stage)
+    {
+      vtkErrorWithObjectMacro(renderer, << "Stage failed to open");
+      return;
+    }
+
+    // TODO: USD bake skinning is not performant
+    // We need to read joints and do the skinning in the shader
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:522]: create an issue ?

**Review by mwestphal — CHANGES_REQUESTED** [review]
only small changes needed

File: plugins/usd/module/vtkF3DUSDImporter.cxx [hunk 24]
```diff
@@ -0,0 +1,968 @@
...
+          polydata = this->MeshMap[meshPrim.GetPath().GetAsString()];
+          bool meshAlreadyExists = (polydata != nullptr);
+
+          if (!meshAlreadyExists)
+          {
+            polydata = vtkSmartPointer<vtkPolyData>::New();
+          }
+
+          
+
```
Comment (mwestphal) [inline plugins/usd/module/vtkF3DUSDImporter.cxx:162]: lots of empty lines

File: plugins/usd/module/vtkF3DUSDImporter.cxx (hunk 23 above)
Comment (mwestphal) [reply plugins/usd/module/vtkF3DUSDImporter.cxx:522]: add link to the issue, here and everywhere needed.

**Review by mwestphal — APPROVED** [review]
small last comments then gtg
